addons/mail/models/mail_message.py

```diff
@@ -150,6 +150,7 @@ def default_get(self, fields):
     starred = fields.Boolean(
         'Starred', compute='_compute_starred', search='_search_starred', compute_sudo=False,
         help='Current user has a starred notification linked to this message')
+    pinned_on_channel_id = fields.Many2one('mail.channel', 'pinned_message_ids')
```

Comment:
As a message is linked to a single record (no more shared messages), I don't think we need yet another foreign key ... a "is_pinned" boolean field is sufficient, you just have to do the inverse o2m based on res_id / res_model normally. This helps keeping the model as lightweight as possible.